sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubWriteSchemaTransformConfiguration.java

```diff
@@ -0,0 +1,205 @@
+ * repository.
+ */
+@Experimental
+@DefaultSchema(AutoValueSchema.class)
+@AutoValue
+public abstract class PubsubWriteSchemaTransformConfiguration {
+
+  public static final String DEFAULT_TIMESTAMP_ATTRIBUTE = "event_timestamp";
+
+  public static Builder builder() {
+    return new AutoValue_PubsubWriteSchemaTransformConfiguration.Builder();
+  }
+
+  public static TargetConfiguration.Builder targetConfigurationBuilder() {
+    return new AutoValue_PubsubWriteSchemaTransformConfiguration_TargetConfiguration.Builder()
+        .setTimestampAttributeKey(DEFAULT_TIMESTAMP_ATTRIBUTE);
+  }
+
+  public static SourceConfiguration.Builder sourceConfigurationBuilder() {
+    return new AutoValue_PubsubWriteSchemaTransformConfiguration_SourceConfiguration.Builder();
+  }
+
+  /**
+   * Configuration details of the source {@link org.apache.beam.sdk.values.Row} {@link
+   * org.apache.beam.sdk.schemas.Schema}.
+   */
+  @Nullable
+  public abstract SourceConfiguration getSource();
+
+  /** Configuration details of the target {@link PubsubMessage}. */
+  public abstract TargetConfiguration getTarget();
+
+  /**
+   * The topic to which to write Pub/Sub messages.
+   *
+   * <p>See {@link PubsubIO.PubsubTopic#fromPath(String)} for more details on the format of the
+   * topic string.
+   */
+  public abstract String getTopic();
+
+  /**
+   * The expected format of the Pub/Sub message.
+   *
+   * <p>Used to retrieve the {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializer} from
+   * {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializers}.
+   */
+  @Nullable
+  public abstract String getFormat();
+
+  /**
+   * When writing to Cloud Pub/Sub where unique record identifiers are provided as Pub/Sub message
+   * attributes, specifies the name of the attribute containing the unique identifier.
+   */
+  @Nullable
+  public abstract String getIdAttribute();
+
+  /** Builder for {@link PubsubWriteSchemaTransformConfiguration}. */
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    /**
+     * Configuration details of the source {@link org.apache.beam.sdk.values.Row} {@link
+     * org.apache.beam.sdk.schemas.Schema}.
+     */
+    public abstract Builder setSource(SourceConfiguration value);
+
+    /** Configuration details of the target {@link PubsubMessage}. */
+    public abstract Builder setTarget(TargetConfiguration value);
+
+    /**
+     * The topic to which to write Pub/Sub messages.
+     *
+     * <p>See {@link PubsubIO.PubsubTopic#fromPath(String)} for more details on the format of the
+     * topic string.
+     */
+    public abstract Builder setTopic(String value);
+
+    /**
+     * The expected format of the Pub/Sub message.
+     *
+     * <p>Used to retrieve the {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializer}
+     * from {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializers}.
+     */
+    public abstract Builder setFormat(String value);
```

Comment:
thanks Damon! perhaps you can list here the formats that are supported? I understand it's anything that's a Payload serializer, but since users may use this without an IDE / writing unit tests, they may not know a priori what those are... wdyt?

Reply:
@pabloem Should the code comment instead look something like this:
```
/**
     * The expected format of the Pub/Sub message.
     *
     * <p>Used to retrieve the {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializer}
     * from {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializers}.  See list of supported format
     * values by invoking {@link org.apache.beam.sdk.schemas.io.Providers#loadProviders}:
     * <pre>{@code Providers.loadProviders(PayloadSerializer.class) 
     * }</pre>.
     */
``

Reply:
I think I meant a list of "JSON", "AVRO" - but I understand that may not be feasible. LMK what you thik.

Reply:
@pabloem I like the idea of explicitly listing.  However, it introduces the coupling of documentation and code.  New implementations of `PayloadSerializer` would have to then add the identifier to this code comment.  I'm on the fence but leaning more towards updating the code comment mentioned above.  I recopied it below.
```java
/**
     * The expected format of the Pub/Sub message.
     *
     * <p>Used to retrieve the {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializer}
     * from {@link org.apache.beam.sdk.schemas.io.payloads.PayloadSerializers}.  See list of supported format
     * values by invoking {@link org.apache.beam.sdk.schemas.io.Providers#loadProviders}:
     * <pre>{@code Providers.loadProviders(PayloadSerializer.class) 
     * }</pre>.
     */
```